### PR TITLE
Add function restoreToCount to Canvas.

### PR DESCRIPTION
**Pull request by guoguo338**
This PR is to fix issue: https://github.com/flutter/flutter/issues/110324

image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/painting/canvas.h [hunk 1]
```diff
@@ -48,6 +48,7 @@ class Canvas : public RefCountedDartWrappable<Canvas>, DisplayListOpFlags {
 
   void restore();
   int getSaveCount();
+  void restoreToCount(size_t count);
```
Comment (ColdPaleLight) [inline lib/ui/painting/canvas.h:51]: Can we just use `int` here?

Comment (guoguo338) [reply lib/ui/painting/canvas.h:51]: yes, done

File: lib/ui/painting.dart [hunk 2]
```diff
@@ -4691,6 +4691,16 @@ class Canvas extends NativeFieldWrapperClass1 {
   @FfiNative<Void Function(Pointer<Void>)>('Canvas::restore', isLeaf: true)
   external void restore();
 
+  /// Pops the save stack until the count layer, if there is anything to pop.
+  /// Otherwise, does nothing.
```
Comment (flar) [inline lib/ui/painting.dart:4695]: Add a comment about how this relates to getSaveCount(). Something along the lines of "the count layer that was previously retrieved with [getSaveCount]" or "After this method completes, the [count] should match the return value of [getSaveCount]".

Also, describe what happens if count > getSaveCount(). I believe it just gets ignored, but we should document that.

**Review by flar — CHANGES_REQUESTED** [review]
Some minor suggestions on the doc comment which I would like to review when done. Otherwise it looks great!

File: lib/ui/painting.dart (hunk 2 above)
Comment (flar) [reply lib/ui/painting.dart:4695]: How about:

```
/// Restores the save stack to a previous level as might be obtained from [getSaveCount].
/// If [count] is less than 1, the stack is restored to its initial state. If [count] is
/// greater than the current [getSaveCount] then nothing happens.
```

I hinted at what happens with a negative count and also implied that the initial save count is 1 (is that true?) - perhaps we should add a test that verifies these behaviors. I believe that one of the web implementations may not be protected against restoring too far which means that behavior will end up blowing up compared to the other implementations (check that restore protects against an empty stack and also that restoreToCount of values less than 1 all behave the same on all platforms).

Comment (guoguo338) [reply lib/ui/painting.dart]: Get that, I have added test to cover those corner cases.

File: lib/ui/painting.dart [hunk 3]
```diff
@@ -4805,6 +4805,17 @@ class Canvas extends NativeFieldWrapperClass1 {
   @FfiNative<Void Function(Pointer<Void>)>('Canvas::restore', isLeaf: true)
   external void restore();
 
+  /// Restores the save stack to a previous level as might be obtained from [getSaveCount].
+  /// If [count] is less than 1, the stack is restored to its initial state.
+  /// If [count] is greater than the current [getSaveCount] then nothing happens.
+  ///
+  /// Use [save] and [saveLayer] to push state onto the stack.
+  ///
+  /// If the state was pushed with [saveLayer], then this call will also
```
Comment (flar) [inline lib/ui/painting.dart:4814]: Nit - this paragraph is worded as if there is only one stack level being restored. Please reword it:

```
/// If any of the state stack levels restored by this call were pushed with
/// [saveLayer], then this call will also cause those layers to be composited
/// into their previous layers.
```

**Review by flar — APPROVED** [review]
Please fix the paragraph mentioned before pushing this change.

File: lib/ui/painting.dart (hunk 3 above)
Comment (guoguo338) [reply lib/ui/painting.dart]: Get, it's done.

**Review by ColdPaleLight — APPROVED** [review]
LGTM!

**Pull request merged**